strawberry/types/fields/field.py

```diff
@@ -0,0 +1,144 @@
+from typing import Callable, Optional, Set, Type, TypeVar, Union
+
+from strawberry.types import StrawberryArgument, StrawberryObject, \
+    StrawberryResolver, StrawberryType
+
+T = TypeVar("T")
+S = TypeVar("S")
+_RESOLVER_TYPE = Union[StrawberryResolver[S], Callable[..., S]]
+
+
+class StrawberryField(StrawberryType[T]):
+    """Strawberry field
+
+    >>> StrawberryField(
+    ...     type_=int,
+    ...     resolver=lambda: 5,
+    ...     name="cool_field"
+    ... )
+    """
+
+    class _NO_DEFAULT_VALUE:
+        """Sentinel value to indicate that a field has no default.
+
+        `None` is not a good option because it can actually be the default for a
+        nullable type.
+        """
+
+    class _NO_FIELD_TYPE:
+        """Sentinel value to indicate a field has not received a type"""
+
+    def __init__(
+        self, *,
+        type_: Optional[Type[T]] = None,
+        # TODO: Should this _only_ take StrawberryResolver, and decorator would
+        #       handle conversion from Callable to StrawberryResolver?
+        resolver: Optional[_RESOLVER_TYPE[T]] = None,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        permission_classes: Optional[object] = None,
+    ):
+
+        # TODO: Move this check to someplace better
+        if (type_ and resolver) and (type_ != resolver.type):
+            # TODO: Raise Strawberry exception
+            ...
+
+        self.resolver = self._convert_resolver(resolver) if resolver else None
+
+        self._type = type_
+        self._name = name
+        self._description = description
+        # TODO: Deal with permissions
+        self.permission_classes = permission_classes
+        # TODO: Default value. Maybe just use a lambda that returns static value
+        #       as resolver instead
+        self.default_value: StrawberryObject = self._NO_DEFAULT_VALUE
+
+    def __call__(self, resolver: _RESOLVER_TYPE[T]) -> None:
+        self.resolver = self._convert_resolver(resolver)
+
+    @property
+    def arguments(self) -> Set[StrawberryArgument]:
+        if self.resolver:
+            return self.resolver.arguments
+
+        # TODO: Should we raise an exception instead?
+        return set()
+
+    @property
+    def description(self) -> Optional[str]:
+        return self._description
+
+    @property
+    def name(self) -> str:
+        """The name specified explicitly, or if not set the wrapped resolver's
+        name
+        """
+        if self._name is not None:
+            return self._name
+
+        if self.resolver is not None:
+            return self.resolver.name
+
+        # TODO: Should we raise an exception instead?
```

Comment:
This case should never happen, so I’d say yes :)

Reply:
I coudl see someone/something trying to access the name of the field, without first setting a resolver. That alone doesn't seem to me like it would warrant an exception

Reply:
what would the usecase be?